src/api/lock.jl

```diff
@@ -0,0 +1,14 @@
+const liblock = ReentrantLock()
+
+# Try to acquire the lock (test-test-set) and close if successful
+# Otherwise, defer finalization
+# https://github.com/JuliaIO/HDF5.jl/issues/1048
+function try_close_finalizer(x)
+    if !islocked(liblock) && trylock(liblock) do
+        close(x)
+        true
+    end
```

Comment:
How does this work? `trylock(f, liblock)` isn't documented, but I assume that inside the `do` block `liblock` will be locked. But doesn't `close` then call an API function which itself will try to get a lock?

Reply:
`trylock(f, liblock)` is just

```julia
function trylock(f, l::AbstractLock)
    if trylock(l)
        try
            return f()
        finally
            unlock(l)
        end
    end
    return false
end
```

You can call `lock` multiple times from the same `Task`, but each `lock` has to be paired with an `unlock`.

```julia
julia> const liblock = ReentrantLock()
ReentrantLock(nothing, 0x00000000, 0x00, Base.GenericCondition{Base.Threads.SpinLock}(Base.InvasiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), (8, 1980324864016, 1980324864048))

julia> lock(liblock)

julia> lock(liblock)

julia> lock(liblock)

julia> islocked(liblock)
true

julia> unlock(liblock)

julia> unlock(liblock)

julia> islocked(liblock)
true

julia> unlock(liblock)

julia> islocked(liblock)
false
```

`trylock` also has to be paired with an `unlock` if successful, but the `do` lambda takes care of that for us.

1. We use `islocked` first because it does not involve atomics and is thus cheaper.
2. `trylock` is only attempted if `islocked` is `false`.

3a. If `trylock` can `lock`, then we enter the `do` block successfully, `close`, and return `true` from the outer `trylock`. 
4a. We then enter the empty `if` block and return `nothing`.

3b. If we cannot `lock`, then we add this to the end of the finalization queue the next time the garbage collection runs.
4b. We return whatever `finalizer` returns, which I think is `x`. Perhaps we should make this type stable and return `nothing`?

Reply:
> You can call lock multiple times from the same Task, but each lock has to be paired with an unlock.

Ah, TIL! That makes sense.

Reply:
> 4b. We return whatever `finalizer` returns, which I think is `x`. Perhaps we should make this type stable and return `nothing`?

That seems like a good idea